File: core/src/main/java/com/starshootercity/OriginsAddon.java
package com.starshootercity;

import com.starshootercity.abilities.Ability;
import com.starshootercity.abilities.AbilityRegister;
import com.starshootercity.events.PlayerSwapOriginEvent;
import com.starshootercity.packetsenders.OriginsRebornResourcePackInfo;
import net.kyori.adventure.key.Key;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.List;

public abstract class OriginsAddon extends JavaPlugin {
    private static OriginsAddon instance;

    public static OriginsAddon getInstance() {
        return instance;
    }

    public @Nullable SwapStateGetter shouldOpenSwapMenu() {
        return null;
    }

    public @Nullable SwapStateGetter shouldAllowOriginSwapCommand() {
        return null;
    }

    public @Nullable KeyStateGetter hasAbilityOverride() {
        return null;
    }

    public interface SwapStateGetter {
        State get(Player player, PlayerSwapOriginEvent.SwapReason reason);
    }

    public interface KeyStateGetter {
        State get(Player player, Key key);
    }

    @SuppressWarnings("unused")
    public enum State {
        ALLOW,
        DEFAULT,
        DENY
    }

    @Override
    public final void onEnable() {
        instance = this;
        onRegister();
        AddonLoader.register(this);
        for (Ability ability : getAbilities()) {
            AbilityRegister.registerAbility(ability, this);
        }
        if (getResourcePackInfo() != null) PackApplier.addResourcePack(this, getResourcePackInfo());
        afterRegister();
    }

    public @Nullable OriginsRebornResourcePackInfo getResourcePackInfo() {
        return null;
    }

    @Override
    public @NotNull File getFile() {
        return super.getFile();
    }

    public void onRegister() {}

    public void afterRegister() {}

    public abstract @NotNull String getNamespace();

    public @NotNull List<Ability> getAbilities() {
        return List.of();
    }
}

File: core/src/main/java/com/starshootercity/OriginsAddon.kt
package com.starshootercity

import com.starshootercity.abilities.Ability
import com.starshootercity.abilities.AbilityRegister
import com.starshootercity.abilities.custom.ToggleableAbility
import com.starshootercity.events.PlayerSwapOriginEvent
import com.starshootercity.packetsenders.OriginsRebornResourcePackInfo
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import java.io.File

abstract class OriginsAddon : JavaPlugin() {

    companion object {
        private lateinit var instance: OriginsAddon
        fun getInstance(): OriginsAddon = instance
    }

    open fun shouldOpenSwapMenu(): SwapStateGetter? = null

    open fun shouldAllowOriginSwapCommand(): SwapStateGetter? = null

    open fun getAbilityOverride(): KeyStateGetter? = null

    interface SwapStateGetter {
        fun get(player: Player, reason: PlayerSwapOriginEvent.SwapReason): State
    }

    interface KeyStateGetter {
        fun get(player: Player, key: Key): State
    }

    @Suppress("unused")
    enum class State {
        ALLOW,
        DEFAULT,
        DENY
    }

    final override fun onEnable() {
        instance = this
        onRegister()
        AddonLoader.register(this)
        for (ability in getAbilities()) {
            if (ability is ToggleableAbility && !ability.shouldRegister()) continue
            AbilityRegister.registerAbility(ability, this)
        }
        getResourcePackInfo()?.let { PackApplier.addResourcePack(this, it) }
        afterRegister()
    }

    open fun getResourcePackInfo(): OriginsRebornResourcePackInfo? = null

    public override fun getFile(): File = super.getFile()

    open fun onRegister() {}

    open fun afterRegister() {}

    abstract fun getNamespace(): String

    open fun getAbilities(): List<Ability> = listOf()
}
